x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java

```diff
@@ -0,0 +1,142 @@
+import org.elasticsearch.xpack.sql.expression.NamedExpression;
+import org.elasticsearch.xpack.sql.expression.function.Function;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+import org.elasticsearch.xpack.sql.tree.Source;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static java.util.Collections.singletonList;
+
+public class Pivot extends UnaryPlan {
+
+    private final Expression column;
+    private final List<NamedExpression> values;
+    private final List<NamedExpression> aggregates;
+    // derived properties
+    private AttributeSet groupingSet;
+    private AttributeSet valueOutput;
+    private List<Attribute> output;
+
+    public Pivot(Source source, LogicalPlan child, Expression column, List<NamedExpression> values, List<NamedExpression> aggregates) {
+        super(source, child);
+        this.column = column;
+        this.values = values;
+        this.aggregates = aggregates;
+    }
+
+    @Override
+    protected NodeInfo<Pivot> info() {
+        return NodeInfo.create(this, Pivot::new, child(), column, values, aggregates);
+    }
+
+    @Override
+    protected Pivot replaceChild(LogicalPlan newChild) {
+        return new Pivot(source(), newChild, column, values, aggregates);
+    }
+
+    public Expression column() {
+        return column;
+    }
+
+    public List<NamedExpression> values() {
+        return values;
+    }
+
+    public List<NamedExpression> aggregates() {
+        return aggregates;
+    }
+    
+    public AttributeSet groupingSet() {
+        if (groupingSet == null) {
+            AttributeSet columnSet = Expressions.references(singletonList(column));
+            // grouping can happen only on "primitive" fields, thus exclude multi-fields or nested docs
+            // the verifier enforces this rule so it does not catches folks by surprise
+            groupingSet = new AttributeSet(Expressions.onlyPrimitiveFieldAttributes(child().output()))
+                    // make sure to have the column as the last entry (helps with translation)
+                    .subtract(columnSet)
+                    .subtract(Expressions.references(aggregates))
+                    .combine(columnSet);
+        }
+        return groupingSet;
+    }
+
+    public AttributeSet valuesOutput() {
+        // TODO: the generated id is a hack since it can clash with other potentially generated ids
+        if (valueOutput == null) {
+            List<Attribute> out = new ArrayList<>(aggregates.size() * values.size());
+            if (aggregates.size() == 1) {
+                NamedExpression agg = aggregates.get(0);
+                for (NamedExpression value : values) {
+                    ExpressionId id = new ExpressionId(agg.id().hashCode() + value.id().hashCode());
+                    out.add(value.toAttribute().withDataType(agg.dataType()).withId(id));
+                }
+            }
+            // for multiple args, concat the function and the value
```

Comment:
Shouldn't we do that also for one aggregate function?

Reply:
No - this follows the Oracle convention which is the only one that supports multiple aggs inside the same pivot. Currently this branch is disabled due to a corner-case optimization that results in missing IDs.